arch/posix/src/support.rs

```diff
@@ -0,0 +1,55 @@
+use nix::sys::signal::{self, SigSet, SigmaskHow};
+
+extern "C" {
+    fn pause() -> isize;
+}
+
+/// Waits for a signal to be received
+pub fn wfi() {
+    let mut current_signals = SigSet::empty();
+    let masked_signals = SigSet::empty();
+    signal::sigprocmask(
+        SigmaskHow::SIG_SETMASK,
+        Some(&masked_signals),
+        Some(&mut current_signals),
+    )
+    .unwrap();
+
+    // println! ("pause");
+    unsafe {
+        pause();
+    }
+    // println! ("resume");
+
+    signal::sigprocmask(SigmaskHow::SIG_SETMASK, Some(&current_signals), None).unwrap();
+}
+
+pub unsafe fn atomic<F, R>(f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    let mut current_signals = SigSet::empty();
+    let masked_signals = SigSet::all();
+
+    // Mask Signals
+    signal::sigprocmask(
+        SigmaskHow::SIG_SETMASK,
+        Some(&masked_signals),
+        Some(&mut current_signals),
+    )
+    .unwrap();
+    // println! ("{:?}", current_signals);
+    let res = f();
+
+    // Unmask Signals
+    signal::sigprocmask(SigmaskHow::SIG_SETMASK, Some(&current_signals), None).unwrap();
+    res
+}
+
+#[inline(always)]
+/// NOP instruction
+pub fn nop() {
+    unsafe {
+        llvm_asm!("nop" :::: "volatile");
```

Comment:
Instead of using assembly couldn't this just do nothing?

Reply:
While it's a bit of a hack, at certain places where we want a few cycles of delay (entering panic comes to mind) we issue calls to `nop` that the kernel expects to actually take real-world time. It probably doesn't matter in practice for the posix platform, but I do think that the `nop` exposed by the arch should be an actual, architectural `nop` rather than something that could be optimized out.

Reply:
Re the question below, I think that `nop` is a valid instruction on RISC-V, ARM, x86, and x64. That's probably enough to effectively be cross platform anyway.

Reply:
Good point. This is probably ok then. It would be better if we could use a POSIX nop though. Ideally this shouldn't be architecture specific at all.

Reply:
Is there such a thing available? Do you have anything particular in mind?

Reply:
Not really unfortunately. I can't think of a POSIX NOP syscall. As Par mentioned a NOP should be fine here.